server/src/main/java/com/linecorp/centraldogma/server/internal/storage/MigratingMetaToDogmaRepositoryService.java

```diff
@@ -0,0 +1,180 @@
+                    Author.SYSTEM, projectName, REPO_DOGMA, RepositoryStatus.READ_ONLY).join();
+            // Wait for the repository status to be updated.
+            Thread.sleep(3000L);
+            logger.info("Dogma repository in the project '{}' is set to read-only.", projectName);
+            try {
+                migrateMetaRepository(project);
+                numMigratedProjects++;
+                logger.info("Meta repository in the project '{}' is migrated to dogma repository.",
+                            projectName);
+            } catch (Throwable t) {
+                logger.warn("Failed to migrate meta repository of {} project. Set back to active.",
+                            projectName, t);
+                metadataService.updateRepositoryStatus(
+                        Author.SYSTEM, projectName, REPO_DOGMA, RepositoryStatus.ACTIVE).join();
+                // Do not continue the migration if the migration fails.
+                throw t;
+            }
+
+            commandExecutor.execute(Command.resetMetaRepository(Author.SYSTEM, projectName)).join();
+            metadataService.updateRepositoryStatus(
+                    Author.SYSTEM, projectName, REPO_DOGMA, RepositoryStatus.ACTIVE).join();
+            logger.info("Dogma repository of {} project is set to active.", projectName);
+        }
+        return numMigratedProjects;
+    }
+
+    private void logMigrationJob(int numMigratedProjects) throws Exception {
+        final Map<String, Object> data =
+                ImmutableMap.of("timestamp", Instant.now(),
+                                "numMigratedProjects", numMigratedProjects);
+        final Change<JsonNode> change = Change.ofJsonUpsert(META_TO_DOGMA_MIGRATION_JOB,
+                                                            Jackson.writeValueAsString(data));
+        final Command<CommitResult> command =
+                Command.push(Author.SYSTEM, INTERNAL_PROJECT_DOGMA,
+                             REPO_DOGMA, Revision.HEAD,
+                             "Migrating meta repository to dogma repository has been done.", "",
+                             Markup.PLAINTEXT, change);
+        executeCommand(command);
+    }
+
+    private void migrateMetaRepository(Project project) throws Exception {
+        final Repository metaRepository = project.repos().get(Project.REPO_META);
+        final Map<String, Entry<?>> entryMap =
+                metaRepository.find(Revision.HEAD, "/credentials/*.json,/repos/**.json").get();
+        final ImmutableList.Builder<Change<?>> builder =
+                ImmutableList.builderWithExpectedSize(entryMap.size() + 1);
+        for (Map.Entry<String, Entry<?>> entry : entryMap.entrySet()) {
+            //noinspection unchecked
+            final Entry<JsonNode> value = (Entry<JsonNode>) entry.getValue();
+            builder.add(Change.ofJsonUpsert(entry.getKey(), value.content()));
+        }
+        builder.add(Change.ofJsonUpsert(META_TO_DOGMA_MIGRATED,
```

Comment:
If this job fails after migrating some projects, those migrated projects start using the `dogma` repository to read and store metadata anyway. It means the metadata may be updated to the `dogma` repository after the failure. 

My concern is that the new data is overwritten by the old data in the `meta` repository if this job runs again. Should we skip the migration for a project if the project was migrated before?